### DISCUSSION TITLE
How can I disable null-ls/none-ls source which is enabled in extras?

### DISCUSSION THREAD

**Discussion by fredrikaverpil** [LazyVim/LazyVim#1538] [Q&A] [answered]
In my case, I'm trying to disable the `goimports-reviser` that is enabled in the go extas here: https://github.com/LazyVim/LazyVim/blob/5eca3dc3bba081fd1baced79af80da2c60307c8a/lua/lazyvim/plugins/extras/lang/go.lua#L95

I've tried the following:

```lua
local null_ls = require("null-ls")

null_ls.disable("goimports-reviser")
null_ls.disable({ name = "goimports-reviser" })
```

But none of those commands disables it. The only thing that works is that I comment out [this](https://github.com/LazyVim/LazyVim/blob/5eca3dc3bba081fd1baced79af80da2c60307c8a/lua/lazyvim/plugins/extras/lang/go.lua#L95) line from my local `lazyvim/plugins/extras/lang/go.lua` file.

This is my none-ls config and I tried the above commands in the `opts` without success: https://github.com/fredrikaverpil/dotfiles/blob/main/nvim-lazyvim/lua/plugins/lsp.lua#L66

Can someone see what I'm doing wrong here?

**Comment by folke** [comment]
goimports was actually removed in a PR today.
I'm pushing a new release in a minute.
https://github.com/LazyVim/LazyVim/pull/1536

**Reply by fredrikaverpil** [reply]
I think that PR might have been made in error... I added [a comment](https://github.com/LazyVim/LazyVim/pull/1535#issuecomment-1741952706) about it in the PR.

Although it was nice to see `goimports-reviser` removed, as I didn't want that, I still feel I must be doing something wrong in my own config if I cannot disable a null-ls formatter that is enabled from a lazyvim language extras...

**Reply by Nikola-Milovic** [reply]
I am out of loop a bit, will `goimports` be added back into `go.extras`? I always had it enabled when I had my own custom setup, `gopls` does not provide that

**Reply by fredrikaverpil** [reply]
@Nikola-Milovic you can just keep `goimports` in your own config, like I do [here](https://github.com/fredrikaverpil/dotfiles/blob/main/nvim-lazyvim/lua/plugins/lsp.lua). Or open a PR to put it back into the go extras I guess.

**Comment by dpetka2001** [comment]
@fredrikaverpil 
I tried what you said but with `prettierd`. I tried the following code 
```lua
    opts = function(_, opts)
      local nls = require("null-ls")
      opts.sources = vim.list_extend(opts.sources, {
        nls.builtins.formatting.prettierd,
      })
	  nls.disable("prettierd")
    end,
```
and also could not disable `prettierd`. But then I tried in a `.html` file from command mode `require("null-ls").disable("prettierd")` and it worked fine. I thought there might be some kind of race condition between setting up the sources and disable what you want. So I wrapped the disable command in a `vim.schedule` function like so 
```lua
      vim.schedule(function()
        nls.disable("prettierd")
      end)
```
and it worked correctly. I understand you might not need it now that the sources have been removed from the Go extra, just thought I'd mention it for future reference and maybe if you also wanted to test out that it works.

**Reply by folke** [reply]
the correct way swould be to iterate over the list of sources and remove the correct one with the given name.

**Reply by dpetka2001** [reply]
Tnx for pointing that out @folke, I just went with the command `disable` that the OP mentioned in his initial post and why that wouldn't work.

**Reply by folke** [reply]
disable doesnt work since null-ls has not been set up yet in that case. That `vim.schedule` is a great solution for sure!

**Reply by dpetka2001** [reply]
Tnx for helping me understand even better :smile:

**Reply by fredrikaverpil** [reply]
Hey, thanks so much @dpetka2001 for taking your time to look into this. I ended up with this solution after having read @folke's input, and it works for me (althought `goimports-reviser` is no longer part of the go extras).

```lua
      local remove_sources = { "goimports_reviser" }
      for _, source in ipairs(remove_sources) do
        for i, v in ipairs(opts.sources) do
          -- vim.api.nvim_echo({ { "Checking " .. v.name, "None" } }, false, {})
          if v.name == source then
            -- vim.api.nvim_echo({ { "Removing " .. source .. " from null-ls sources", "None" } }, false, {})
            table.remove(opts.sources, i)
          end
        end
      end
```

**Reply by folke** [reply, marked as answer]
Not sure if that will work with more than one source to remove, since you're changing the table you are iterating over.

Something like this is probably safer:
```lua
local remove_sources = { "goimports_reviser" }
opts.sources = vim.tbl_filter(function(source)
  return not vim.tbl_contains(remove_sources, source.name)
end, opts.sources)
```

**Reply by fredrikaverpil** [reply]
🤦 ah yes, thanks for noticing that mistake.
